### PR TITLE
Refactored dataset-delete API to DELETE method with URI which is  /datasets/{dataset}

### PR DESCRIPTION
**Pull request by siddardh-ra**
Refactored `datasets/delete` to be a `DELETE` operation from the `POST` method, which will look like  `DELETE /dataset/<resource_id>`

### CODE REVIEW

File: dashboard/src/actions/overviewActions.js
Comment (dbutenhof) [inline dashboard/src/actions/overviewActions.js:218]: I liked `update`, although I suppose `modify` works as well. Why the change? I suppose you could argue that *neither* really clean encompasses "delete"; but I certainly don't think "modify" does any better.

We're gradually integrating this into an all-purpose "resource" API, which could also take over some evolution of `datasets_detail` as well, and might well take over the `datasets_metadata` `GET`/`PUT` APIs. That might argue for making this endpoint label simply `datasets`.

E.g.,
* `GET /datasets` would be the new `datasets_list`,
* `GET /datasets/{id}` would integrate `datasets_detail` and the `GET` on `datasets_metadata` with query parameters to specify a "view", `?metadata` to get metadata, etc.
* `PUT /datasets/{id}` updates METADATA via the JSON `{"metadata":{...}}` body (although it might be more "RESTful" to integrate this as a resource *action* under `POST` since we wouldn't be treating `metadata` as a separate sub-resource as `/datasets/metadata` implies)
* `POST /datasets/{id}` allows publish, "re-own" and maybe others, via query parameters ...
* `DELETE /datasets/{id}` allows deleting the resource.

File: lib/pbench/server/api/resources/query_apis/__init__.py
Comment (dbutenhof) [inline lib/pbench/server/api/resources/query_apis/__init__.py:835]: I think that `_post` and `_delete` could both be dependent on a new common `ElasticBulkBase` method with all the surrounding boilerplate instead of duplicating it. Also, each subclass has its own `action` for the generate actions helper, and it's true that delete doesn't need the `"doc"`, but I think the common base class can still just call `self.generate_actions` and the specifics can be handled by the generator based on the object context. For example, we'd talked earlier about using the `Schema` for this, with an `attributes` parameter that would be carried through and allow the `DELETE` schema to say it wants the delete action and no doc where the `POST` schema wants update and triggers building the appropriate doc.

This structure would be cleaner and easier to maintain than the current duplication.

File: lib/pbench/server/api/resources/query_apis/datasets_modify.py
Comment (dbutenhof) [inline lib/pbench/server/api/resources/query_apis/datasets_modify.py:57]: `attributes={"action": "delete", "build_doc": False}`, for example? Where the `Schema.__init__` would simply record `self.attributes`, the new `_post`/`_delete` common code method would attach the schema attributes to the API `context`, and the action builder would read the attributes to "do the right thing"... ??

File: lib/pbench/test/unit/server/query_apis/test_datasets_modify.py
Comment (dbutenhof) [inline lib/pbench/test/unit/server/query_apis/test_datasets_modify.py:297]: This looks like you just combined the two unit test modules.

First, you don't have to do that, because there's no rule that each endpoint can only be tested by a single pytest file. It's fine to have a file for delete and a file for "modify" if that's convenient. However, you could also combine the two *test *classes** within the same file, and still have them operate independently so you don't have to worry about defining two `test_partial` methods and such on the same class...

**Review by dbutenhof — CHANGES_REQUESTED** [review]
Great start, but I think it can be improved by some refactoring!

File: dashboard/src/actions/overviewActions.js
Comment (webbnh) [reply dashboard/src/actions/overviewActions.js:218]: I, too, have a mild preference for `update` over `modify`, but I think I like the more resource-oriented API which Dave is proposing here.

I'm assuming that most of those changes should be dealt with in separate PRs...that _this_ PR will be limited to just `DELETE`.  But, it would be good if this PR laid the groundwork for those as follow-ons.

(However, I think I still prefer `metadata` as a sub-resource of `dataset/{id}`, but that's a conversation for a different time and place.)

**Comment by webbnh** [comment]
Given the level of change which @dbutenhof is requesting (and which I support), I'm going to hold off reviewing these changes until that lands.

File: dashboard/src/actions/overviewActions.js
Comment (dbutenhof) [reply dashboard/src/actions/overviewActions.js:218]: > (However, I think I still prefer `metadata` as a sub-resource of `dataset/{id}`, but that's a conversation for a different time and place.)

Yeah, I'm on the fence about that. Metadata is an aspect of a dataset, and treating it as direct attributes is fine; but thinking of metadata keys as a sub-resource doesn't really bother me, either... and that might argue for `/datasets/{id}/metadata`.

But to re-emphasize what we have both said here, *this* PR shouldn't go beyond integrating `DELETE`.

File: lib/pbench/server/api/resources/query_apis/dataset.py
Comment (dbutenhof) [inline lib/pbench/server/api/resources/query_apis/dataset.py:186]: Looking at the restructured code, I realize there's something I missed. A failed `DELETE` will leave the delete operation in `WORKING` state. We don't need to update anything on *success* because the dataset is gone; but on failure we should really do a `sync.error(message)` ... even though the common `message` refers to `update`, that's probably fine... although that could easily be fixed by using `f"Unable to {action} some indexed documents"` up at line 167.

**Review by dbutenhof — CHANGES_REQUESTED** [review]

**Review by webbnh — CHANGES_REQUESTED** [review]
Looks pretty good, as far as I can tell; however, I found a bunch of small stuff which could use polishing, and I have a few questions.

File: lib/pbench/server/api/resources/query_apis/dataset.py
Comment (dbutenhof) [reply lib/pbench/server/api/resources/query_apis/dataset.py:155]: The new operational sync mechanism doesn't depend on the previous state like the old `dataset.state` did. Therefore it's perfectly OK to retry from `WORKING` or `FAILED`. The comment below was that we really should set to `FAILED` on an error, with `sync.error` and a reasonable message.

In reference to this comment, maybe just drop everything from the ";" on.

File: lib/pbench/server/api/resources/query_apis/dataset.py
Comment (webbnh) [reply lib/pbench/server/api/resources/query_apis/dataset.py:75]: > I'm wondering if we should completely remove the "require_map" now.

You've convinced _me_ that we should....  😁

File: lib/pbench/server/api/resources/query_apis/dataset.py
Comment (siddardh-ra) [reply lib/pbench/server/api/resources/query_apis/dataset.py:30]: Yeah @webbnh. Initially, I thought the same, but a directory named `datasets` is present there. So If I call it `datasets`, then by default it will point to the file `datasets/__init__.py`.

Comment (webbnh) [reply lib/pbench/server/api/resources/query_apis/dataset.py:30]: That's an interesting dilemma, @siddardh-ra.  I presume that you are right that we cannot name this module `datasets.py` if there is already a `datasets` directory.  However, given our existing directory structure (and API name space approach) and how Python maps modules to filenames, perhaps it would be appropriate to put this code in `datasets/__init__.py`.

That is, from a Python perspective, this likely is the fundamental class in the `pbench.server.api.resources.query_apis.datasets` package, so it would be entirely reasonable for its code to be found in the `__init__.py` module.  That is how single-module packages are arranged, and as additional modules added to the package, their files are added to the directory.  In our case, we're kind of backing into this, where have multiple modules and we are contracting the interface, but the result should be the same:  as we convert sibling APIs into separate methods on the same resource, it would make sense that they be moved into the central module.

File: lib/pbench/server/api/resources/query_apis/dataset.py
Comment (webbnh) [reply lib/pbench/server/api/resources/query_apis/dataset.py]: > it's perfectly OK to retry from `WORKING` or `FAILED`.

@dbutenhof, retrying from `FAILED` makes sense, but...if the dataset is in `WORKING` state, mightn't it be the case that some other request is still in the middle of processing (and not yet `FAILED` or `OK`)?  If that were the case, do we really want to proceed?  (I suppose we could just let the ES server sort out the conflicts, if there are any....)

(In any case, the way the code is structured now, on the failure path we'll call `Sync.error()` which will mark it `FAILED`, so at least that loose end is tied off.)

**Review by webbnh — CHANGES_REQUESTED** [review]
Did you confuse `OperationName` with `OperationCode`?

File: lib/pbench/server/api/resources/query_apis/dataset.py
Comment (dbutenhof) [reply lib/pbench/server/api/resources/query_apis/dataset.py:30]: You're not entirely wrong, but "it's complicated". Both `IndexMapBase`, which lives in `pbench.server.api.resource.query_apis.datasets.__init__.py` and `ElasticBulkBase` depend on the index map, but they focus on different aspects. They *could* both extend a common index-map helper, and if `ElasticBulkBase` moved into the subdirectory then so would `query_apis.dataset.py` and `Datasets`.

That's a lot more refactoring than I'd like to see in this PR, however.

File: lib/pbench/server/api/resources/__init__.py
Comment (dbutenhof) [reply lib/pbench/server/api/resources/__init__.py:763]: No, apparently: this is the `Sync` operation (`OperationName`) associated with the API, whereas `OperationCode` is the CRUD operation.

**Review by riya-17 — APPROVED** [review]

File: lib/pbench/server/api/resources/__init__.py
Comment (dbutenhof) [inline lib/pbench/server/api/resources/__init__.py:759]: As a side note, here, this is a great use for the `@dataclass` decorator. Instead of writing an `__init__` constructor that explicitly copies parameters to identically named attributes, you just define the attributes and the dataclass infrastructure does the rest:

```
@dataclass
class ApiAttributes:
    action: str
    operation_name: OperationName
    require_stable: bool
    require_map: bool
```

File: lib/pbench/server/api/resources/query_apis/__init__.py
Comment (dbutenhof) [inline lib/pbench/server/api/resources/query_apis/__init__.py:764]: I think you should avoid this and simply reference `context["attributes"]` directly, for example at line 781, `if context["attributes"].require_stable and operation`.

I think that any `ElasticBulkBase` API that can get here will have `context["attributes"]` ... which is somewhat illustrated already by this code since you're leaving `self.require_stable` *undefined* by default, which would cause an exception below. You could enforce that in the constructor, "to be safe".

**Review by dbutenhof — DISMISSED** [review]
I have a couple of comments; both would slightly simplify the code, but I'll approve anyway and let you decide whether to pursue them.

**Review by webbnh — DISMISSED** [review]
I guess that there is nothing left for which I would gate this PR, but I found a couple of editorial things in this last update which you should probably address, and I have a concern about the `ElasticBulkBase` instance attributes vs. the contents of the `generate_actions()` `context` which I think you should consider reconciling before merging.

Also, I have two open conversations which I'd like to close:
- [ ] If you could accept or decline [this change](https://github.com/distributed-system-analysis/pbench/pull/3252#discussion_r1113546975).
- [ ] If you would respond to [Dave's suggestion](https://github.com/distributed-system-analysis/pbench/pull/3252#discussion_r1114281199).

And, I second [Dave's suggestion](https://github.com/distributed-system-analysis/pbench/pull/3252#discussion_r1121698045) to use an `@dataclass` for `ApiAttributes`.

File: lib/pbench/server/api/resources/query_apis/dataset.py
Comment (siddardh-ra) [reply lib/pbench/server/api/resources/query_apis/dataset.py]: I'll address the removal of `require_map` completely in a  separate PR

File: lib/pbench/server/api/resources/query_apis/__init__.py
Comment (dbutenhof) [inline lib/pbench/server/api/resources/query_apis/__init__.py:710]: 👍🏻

**Review by dbutenhof — APPROVED** [review]

**Review by webbnh — DISMISSED** [review]
LGTM, however, your test run failed due to a confluence of issues which #3270 should fix.

So, one way forward is to wait until that's merged, rebase on it, and then see if the tests proceed.  (Another way is to rebase on it before it's merged....  And, another is for someone to manually fix up your Jenkins workspace.)

**Comment by webbnh** [comment]
> one way forward is to wait until that's merged, rebase on it, and then see if the tests proceed

#3270 is now merged, so you should rebase your branch and see if the tests now pass.

**Review by dbutenhof — CHANGES_REQUESTED** [review]
Oops! Sorry, Siddardh! When I fixed the alembic `UPDATE` enum, I added a new functional test to publish/unpublish datasets, and you've now rebased on top of that. The test is failing because the new `PbenchServerClient.update` method references `API.DATASETS_UPDATE`; you'll need to change that method!

That's in `lib/pbench/client/__init__.py` ...

**Review by dbutenhof — APPROVED** [review]

**Comment by dbutenhof** [comment]
Yeah; we need #3318 ...

> AssertionError: Register failed with {'message': 'Invalid email: functional@gmail.com'}
assert 400 == <HTTPStatus.CREATED: 201>
>  +  where 400 = <Response [400]>.status_code
>  +  and   <HTTPStatus.CREATED: 201> = HTTPStatus.CREATED

**Review by webbnh — APPROVED** [review]
👍

**Pull request merged**
